Real-Time AI Surgical Video Processing Workflow

bhashemian opened the pull request:
This MR:
- Makes necessary changes to enable workflow concept in holohub.
- Add Real-Time End-to-end AI Surgical Video Workflow,
- Implements a lazy import to enable workflows to run without HSB if needed.


Note: This workflow depends on Holoscan SDK v3.0 for Dynamic Flow Condition.

jjomier commented on applications/endoscopy_out_of_body_detection/CMakeLists.txt:
I don't think we should modify this application

jjomier commented on applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.py:
I don't think we should modify this application

bhashemian replied:
You are right, @jjomier! These changes are comming from my last exploration for the surgical tool tracking to work. I will revert them.

jjomier commented on applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.yaml:
I don't think we should modify this application

jjomier commented on dev_container:
We should not modify the base image - this can be passed via command line

jjomier commented on run:
```diff
@@ -114,21 +114,31 @@ get_buildtype_str() {
 get_app_source_root_dir() {
   # Get source path for a given application according to HoloHub convention.
-  # Usage: ./run get_app_source_root_dir <app_name>
+  # Usage: ./run get_app_source_root_dir <app_name> [is_workflow]
```
We should change the logic to look first for apps and then workflow directory.

bhashemian replied:
Here I assumed that a workflow is an application (end-to-end application) but just stored under different directory. I will revisit it after making sure that the workflow itself is working with HSB.

tbirdso replied:
Let's avoid updating CLIs for workflows specifically if possible. Workflows are part of a class of runnable projects including "applications" and "benchmarks" as well. Could instead:
- create an `is_workflow()` command to check a unique project name against a globbed list of workflow names
- check for workflows as a fallback like is done for benchmarks below

Impact: Adding workflow-specific args to commands is not scalable, could be confusing or tenuous for devs working on other projects
Risk: Not blocking, but could result in CLI differences if we refactor during the move to `holohub` CLI

bhashemian replied:
Tanks for the comment! I made the changes.

jjomier commented on workflows/metadata.schema.json:
```diff
@@ -0,0 +1,83 @@
+  "$id": "holohub/application/v1",
+  "type": "object",
+  "properties": {
+    "application": {
```
workflow?

bhashemian replied:
You are right but for now, I will keep it as applications. It has more consequences in other part of the code that I need to take into account.

jjomier replied:
Fixed

jjomier commented on utilities/update_readme.sh:
We can actually remove this file, it is not used anymore

jjomier commented on workflows/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+# limitations under the License.
+
+# Make sure to order by alphabetical list
+add_holohub_application(real_time_ai_surgical_video_processing DEPENDS OPERATORS aja_source)
```
Are we using AJA?

bhashemian replied:
Only if the user wants to use AJA as source, right now we have three options: Replayer, AJA, and HSB.

jjomier replied:
Sounds good.

jjomier commented on workflows/real_time_ai_surgical_video_processing/CMakeLists.txt:
```diff
@@ -0,0 +1,80 @@
+cmake_minimum_required(VERSION 3.20)
+project(real_time_ai_surgical_video_processing_workflow LANGUAGES NONE)
+
+# Create a custom target that runs the Python script
+add_custom_target(real_time_ai_surgical_video_processing
+    COMMAND Python3::Interpreter ${CMAKE_CURRENT_SOURCE_DIR}/python/main.py
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+)
+
+# Make sure Python interpreter is found
+find_package(Python3 COMPONENTS Interpreter REQUIRED)
+
```
Why do we need this?

bhashemian replied:
@jjomier doesn't it essentially check if the python exisits? Do you think it is redundant?

jjomier replied:
I just don't understand the target.

bhashemian replied:
Oh right! I thought you were taking about the find_package part. I added that before talking to you for enabling the data downloads, which didn't work. Thanks for spotting it, I remove it.

jjomier commented on workflows/real_time_ai_surgical_video_processing/python/main.py:
We should have a specific name for the python script not just main.py

bhashemian replied:
I see your point! I thought having all the main python codes as main.py would make it easier for the user to run the workflow and applications, so it always is like:

```bash
python workflows/<WORKFLOW NAME>/python/main.py
```

What do you think?

bhashemian replied:
changed it to `real_time_ai_surgical_video_processing.py`

jjomier commented:
There are linting issues as well as metadata validity.

jjomier commented:
Also linting and metadata issues from the CI to be fixed.

jjomier commented on workflows/real_time_ai_surgical_video_processing/python/main.py:
We should rename main.py to a more explicit name

bhashemian replied:
changed it to real_time_ai_surgical_video_processing.py

jjomier commented on workflows/real_time_ai_surgical_video_processing/README.md:
````diff
@@ -0,0 +1,130 @@
+The repo level build command
+
+```sh
+./run build real_time_ai_surgical_video_processing
````
We should recommend the build_and_run command instead

bhashemian replied:
I included both.

jjomier commented on workflows/real_time_ai_surgical_video_processing/README.md:
```diff
@@ -0,0 +1,152 @@
+
+This workflow utilizes three AI models:
+
+- [Out-of-body detection model](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/endoscopy_out_of_body_detection): `out_of_body_detection.onnx`
+- [SSD model from NGC](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/ssd_surgical_tool_detection_model) with additional NMS op: `epoch24_nms.onnx`
+- [MONAI tool segmentation model from NGC](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/monai_endoscopic_tool_segmentation_model): `model_endoscopic_tool_seg_sanitized_nhwc_in_nchw_out.onnx`
```
```suggestion
- [Out-of-body detection model](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/endoscopy_out_of_body_detection): `out_of_body_detection.onnx`
- [SSD model](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/ssd_surgical_tool_detection_model) with additional NMS op: `epoch24_nms.onnx`
- [MONAI tool segmentation model](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/monai_endoscopic_tool_segmentation_model): `model_endoscopic_tool_seg_sanitized_nhwc_in_nchw_out.onnx`
```

jjomier commented on workflows/real_time_ai_surgical_video_processing/README.md:
```diff
@@ -0,0 +1,152 @@
+
+[📦️ (NGC) Sample App Data for AI-based Endoscopy Tool Tracking](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/holoscan_endoscopy_sample_data)
+
+[📦️ (NGC) Sample App Data for Out-of-Body Detection](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/holoscan_endoscopy_sample_data)
```
The data should be changed

jjomier commented on workflows/real_time_ai_surgical_video_processing/README.md:
```diff
@@ -0,0 +1,152 @@
+   - SSD detection for surgical tool detection
+   - MONAI segmentation for endoscopic tool segmentation
+
+![Sample Endoscopy Image](images/RAISVP-sample-image.png)
```
Is the image still correct?

jjomier commented on workflows/real_time_ai_surgical_video_processing/README.md:
```diff
@@ -0,0 +1,152 @@
+
+[📦️ (NGC) Sample App Data for Out-of-Body Detection](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/holoscan_endoscopy_sample_data)
+
+## Requirements
```
Let's start the README with a quick start using the `build_and_run` command.
I would limit the workflow to build_and_run for maintenance purposes.

bhashemian replied:
Thanks for this feedback! I made a "Quick start guide" section that uses only `build_and_run`, and moved the rest to "Advanced usage" section

tbirdso commented on workflows/real_time_ai_surgical_video_processing/python/real_time_ai_surgical_video_processing.py:
```diff
@@ -0,0 +1,856 @@
+        else:
+            self.add_flow(replayer, source)
+        # __________________________________________________________________
+        # Main Branch
```
I like your comments here, nice for readability!

tbirdso reviewed (commented):
Looking good! Please add a note somewhere, either in the README or 
 a new `workflows/README.md`, with a short overview of how we are defining a "workflow" distinct from other HoloHub project types. The [landing page description](https://nvidia-holoscan.github.io/holohub/) has:

> Reference workflows demonstrate how capabilities from applications and operators can be combined to achieve complex tasks.

Ideally we should add a short docs snippet helping a new developer decide whether their intended use case is a "workflow" or some other project type.

bhashemian commented on workflows/metadata.schema.json:
```diff
@@ -0,0 +1,87 @@
+        "holoscan_sdk": {
+          "$ref": "holohub/project/v1#/$defs/sdk_version"
+        },
+        "holoscscan_sensor_bridge": {
```
Good catch, thanks!

bhashemian commented on workflows/real_time_ai_surgical_video_processing/README.md:
````diff
@@ -0,0 +1,152 @@
+
+Once you have built the Holoscan Sensor Bridge container, you can build the holohub container using the following command:
+
+```sh
+./dev_container build --base_img hololink-demo:2.0.0 --img holohub:link
+```
````
Thanks, we have already provided the one-line command with `build_and_run`. I will remove this container building based on Julien's comment.

bhashemian replied:
Updated the readme.

bhashemian commented on workflows/real_time_ai_surgical_video_processing/README.md:
````diff
@@ -0,0 +1,152 @@
+./run launch real_time_surgical_edge_ai python
+```
+
+Alternatively, you can run the application directly:
````
That's a great point! However, I will remove this to limit the run to build_and_run command based on Julien's comment. Thanks

bhashemian commented on workflows/real_time_ai_surgical_video_processing/README.md:
````diff
@@ -0,0 +1,152 @@
+To run the Python application, you can make use of the run script
+
+```sh
+./run launch real_time_surgical_edge_ai python
````
will remove it, thanks!

bhashemian replied:
Done!

bhashemian commented on workflows/real_time_ai_surgical_video_processing/python/real_time_ai_surgical_video_processing.py:
```diff
@@ -0,0 +1,856 @@
+        spec.output("out")
+
+    def compute(self, op_input, op_output, context):
+        in_messages = op_input.receive("in")
```
They are great questions, @tbirdso! Maybe @gigony or @grlee77 can help us with them.

bhashemian commented on workflows/real_time_ai_surgical_video_processing/python/real_time_ai_surgical_video_processing.py:
```diff
@@ -0,0 +1,856 @@
+        op_output.emit(out_message, "out")
+
+
+class RealTimeAISurgicalVideoProcessingWorkflow(Application):
```
Added a description. Thanks

bhashemian commented on workflows/real_time_ai_surgical_video_processing/README.md:
```diff
@@ -0,0 +1,152 @@
+Fig. 1 Endoscopy (laparoscopy) image from a cholecystectomy (gallbladder removal surgery) showing tool detection and segmentation results from two concurrently executed AI models.
+Image courtesy of Research Group Camma, IHU Strasbourg and the University of Strasbourg ([NGC Resource](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/resources/holoscan_endoscopy_sample_data))
+
+Please refer to the README under [./app_dev_process](./app_dev_process/README.md) to see the process of developing the applications.
```
Good catch! This line was not needed anymore, so removed it.

bhashemian commented on workflows/real_time_ai_surgical_video_processing/README.md:
````diff
@@ -0,0 +1,152 @@
+./docker/build.sh
+```
+
+Once you have built the Holoscan Sensor Bridge container, you can build the holohub container using the following command:
````
Absolutely! Changed them all.

tbirdso commented on workflows/metadata.schema.json:
```diff
@@ -0,0 +1,87 @@
+        "platforms",
+        "tags",
+        "holoscan_sdk",
+        "holoscan_sensor_bridge",
```
HSB should not be required across workflows, please remove

bhashemian commented on workflows/real_time_ai_surgical_video_processing/python/real_time_ai_surgical_video_processing.py:
```diff
@@ -0,0 +1,856 @@
+        op_output.emit(out_message, "out")
+
+
+class ForwardOp(Operator):
```
Sure, I extended the description.

tbirdso commented:
Hi @bhashemian , we have updated HoloHub for Holoscan SDK 3.0.0 base containers. Please rebase your PR and start from a clean build folder to validate. Thanks

bhashemian commented:
@jjomier @tbirdso thanks for much for great comments. I have addressed them all and updated the branch with the latest commit on main. The only remaining item is changing the name of the workflow that I am waining on the final decision. Please let me know if you have any further comments/questions.

tbirdso reviewed (approved):
Looks good to me, nice work @bhashemian !

jjomier reviewed (approved):
LGTM! thanks for the contribution!
